probe-rs/src/flashing/flasher.rs

```diff
@@ -327,11 +332,93 @@ impl<'session> Flasher<'session> {
+    pub(super) fn verify(
+        &mut self,
+        layout: &FlashLayout,
+        ignore_filled: bool,
+    ) -> Result<bool, FlashError> {
         self.run_verify(|active| {
-            active
-                .core
-                .read(fill.address(), page_slice)
-                .map_err(FlashError::Core)
+            if let Some(verify) = active.flash_algorithm.pc_verify {
+                tracing::debug!("Verify using CMSIS function");
+                // Prefer Verify as we may use compression
+                // FIXME: avoid compressing multiple times
+                let flash_encoder =
+                    FlashEncoder::new(active.flash_algorithm.transfer_encoding, layout.clone());
+
+                for page in flash_encoder.pages() {
+                    let address = page.address();
+                    let bytes = page.data();
+
+                    tracing::debug!(
+                        "Verifying page at address {:#010x} with size: {}",
+                        address,
+                        bytes.len()
+                    );
+
+                    // Transfer the bytes to RAM.
+                    let buffer_address = active.load_page_buffer(bytes, 0)?;
+
+                    let result = active.call_function_and_wait(
+                        &Registers {
+                            pc: into_reg(verify)?,
+                            r0: Some(into_reg(address)?),
+                            r1: Some(into_reg(bytes.len() as u64)?),
+                            r2: Some(into_reg(buffer_address)?),
+                            r3: None,
+                        },
+                        false,
+                        Duration::from_secs(30),
+                    )?;
+
+                    // Returns
+                    // status information:
+                    // the sum of (adr+sz) - on success.
+                    // any other number - on failure, and represents the failing address.
+                    if result as u64 != address + bytes.len() as u64 {
+                        tracing::debug!("Verification failed for page at address {:#010x}", result);
+                        return Ok(false);
+                    }
+                }
+            } else {
+                tracing::debug!("Verify using manual comparison");
+                for (idx, page) in layout.pages.iter().enumerate() {
+                    let address = page.address();
+                    let data = page.data();
+
+                    let mut read_back = vec![0; data.len()];
+                    active.read_flash(address, &mut read_back)?;
+
+                    if ignore_filled {
+                        // "Unfill" fill regions. These don't get flashed, so their contents are
+                        // allowed to differ. We mask these bytes with default flash content here,
+                        // just for the verification process.
+                        for fill in layout.fills() {
```

Comment:
We could instead try and ignore the fill ranges. I'm just not sure if doing that in a very general way would be simpler than masking off the fill areas.

Reply:
I think it's fine as it is :)